In [4]:
from gurobipy import *

# Question1

In [58]:
m = Model('question1')

In [59]:
x1 = m.addVar(name='x1')
x2 = m.addVar(name='x2')

In [60]:
m.setObjective(4*x1 + 2*x2, GRB.MAXIMIZE)

In [61]:
c1 = m.addConstr(2*x1 + 4*x2 <= 20, name="c1")
c2 = m.addConstr(3*x1 + 5*x2 <= 15, name="c2")

In [62]:
m.write('question1.lp')

In [63]:
m.optimize()

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 2 rows, 2 columns and 4 nonzeros
Model fingerprint: 0xae31c652
Coefficient statistics:
  Matrix range     [2e+00, 5e+00]
  Objective range  [2e+00, 4e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 2e+01]
Presolve time: 0.00s
Presolved: 2 rows, 2 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.0000000e+31   3.375000e+30   1.000000e+01      0s
       2    2.0000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.01 seconds (0.00 work units)
Optimal objective  2.000000000e+01


In [64]:
for v in m.getVars():
    print(v.varName, v.x)
print('Obj: ' + str(m.objVal))

x1 5.0
x2 0.0
Obj: 20.0


## A. What range of values can the objective function coefficient variable X1 assume without changing the optimal solution? 

In [65]:
print('Sensitivity Analysis (SA) \n ObjVal =', m.ObjVal)
m.printAttr(['X', 'RC', 'Obj', 'SAObjLow', 'SAObjUp'])
m.printAttr(['Sense', 'Slack', 'Pi', 'RHS', 'SARHSLow', 'SARHSUp'])
# Note: printAttr prints only rows with at least one NON-ZERO value, e.g. m.printAttr('X') prints only non-zero variable values.

Sensitivity Analysis (SA) 
 ObjVal = 20.0

    Variable            X           RC          Obj     SAObjLow      SAObjUp 
-----------------------------------------------------------------------------
          x1            5            0            4          1.2          inf 
          x2            0     -4.66667            2         -inf      6.66667 

  Constraint        Sense        Slack           Pi          RHS     SARHSLow      SARHSUp 
------------------------------------------------------------------------------------------
          c1            <           10            0           20           10          inf 
          c2            <            0      1.33333           15            0           30 


The range of value of x1 is [1.2 , inf] assume without change the optimal solution

## B. Is the optimal solution to this problem unique, or are there alternate optimal solutions? 

 The variables x2 have non-zero reduced costs (RC column), which suggests that the current solution is not optimal. A reduced cost of zero indicates that the corresponding variable is at its optimal value in the current solution.

## C.  How much does the objective function coefficient for variable X2 have to increase before it enters the optimal solution at a strictly positive level?

The reduced cost for x2 is -4.66667. This negative reduced cost indicates that the coefficient of x2 in the objective function must increase by at least 4.66667 to reach zero, at which point x2 will enter the solution at a strictly positive level.

## D.  What is the optimal objective function value if X2 equals 1?

In [66]:
m = Model('question1')
x1 = m.addVar(name='x1')
x2 = m.addVar(name='x2')
m.setObjective(4*x1 + 2*(1), GRB.MAXIMIZE)

c1 = m.addConstr(2*x1 + 4*(1) <= 20, name="c1")
c2 = m.addConstr(3*x1 + 5*(1) <= 15, name="c2")
m.optimize()

for v in m.getVars():
    print(v.varName, v.x)
print('Obj: ' + str(m.objVal))

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 2 rows, 2 columns and 2 nonzeros
Model fingerprint: 0x0e1b7870
Coefficient statistics:
  Matrix range     [2e+00, 3e+00]
  Objective range  [4e+00, 4e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+01, 2e+01]
Presolve removed 2 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.5333333e+01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.533333333e+01
x1 3.3333333333333335
x2 0.0
Obj: 15.333333333333334


When I substitute the 1 into the x2, the optimal objective function value is 15.333333333333334 with x1 =  3.333

## E. What is the optimal objective function value if the RHS value for the second constraint changes from 15 to 25? 

In [67]:
m = Model('question1')
x1 = m.addVar(name='x1')
x2 = m.addVar(name='x2')
m.setObjective(4*x1 + 2*x2, GRB.MAXIMIZE)

c1 = m.addConstr(2*x1 + 4*x2 <= 20, name="c1")
c2 = m.addConstr(3*x1 + 5*x2 <= 25, name="c2")
m.optimize()

for v in m.getVars():
    print(v.varName, v.x)
print('Obj: ' + str(m.objVal))

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 2 rows, 2 columns and 4 nonzeros
Model fingerprint: 0xf4f0ef4d
Coefficient statistics:
  Matrix range     [2e+00, 5e+00]
  Objective range  [2e+00, 4e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 3e+01]
Presolve time: 0.00s
Presolved: 2 rows, 2 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.0000000e+31   3.375000e+30   1.000000e+01      0s
       2    3.3333333e+01   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.01 seconds (0.00 work units)
Optimal objective  3.333333333e+01
x1 8.333333333333334
x2 0.0
Obj: 33.333333333333336


The optimal solution is equal to 33.3333 when RHS of second constraint equals to 25

## F. Is the current solution still optimal if the coefficient for the X2 in the second constraint changes from 5 to 1? Explain.

In [73]:
m = Model('question1')
x1 = m.addVar(name='x1')
x2 = m.addVar(name='x2')
m.setObjective(4*x1 + 2*x2, GRB.MAXIMIZE)

c1 = m.addConstr(2*x1 + 4*x2 <= 20, name="c1")
c2 = m.addConstr(3*x1 + 1*x2 <= 15, name="c2")
m.optimize()

for v in m.getVars():
    print(v.varName, v.x)
print('Obj: ' + str(m.objVal))

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 2 rows, 2 columns and 4 nonzeros
Model fingerprint: 0x8aead384
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [2e+00, 4e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 2e+01]
Presolve time: 0.00s
Presolved: 2 rows, 2 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    6.0000000e+30   2.500000e+30   6.000000e+00      0s
       2    2.2000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.00 seconds (0.00 work units)
Optimal objective  2.200000000e+01
x1 4.0
x2 3.0
Obj: 22.0


In [74]:
print('Sensitivity Analysis (SA) \n ObjVal =', m.ObjVal)
m.printAttr(['X', 'RC', 'Obj', 'SAObjLow', 'SAObjUp'])
m.printAttr(['Sense', 'Slack', 'Pi', 'RHS', 'SARHSLow', 'SARHSUp'])
# Note: printAttr prints only rows with at least one NON-ZERO value, e.g. m.printAttr('X') prints only non-zero variable values.

Sensitivity Analysis (SA) 
 ObjVal = 22.0

    Variable            X           RC          Obj     SAObjLow      SAObjUp 
-----------------------------------------------------------------------------
          x1            4            0            4            1            6 
          x2            3            0            2      1.33333            8 

  Constraint        Sense        Slack           Pi          RHS     SARHSLow      SARHSUp 
------------------------------------------------------------------------------------------
          c1            <            0          0.2           20           10           60 
          c2            <            0          1.2           15            5           30 


The optimal of original which equal to 20 has being change to 22 by coefficient for the X2 in the second constraint changes from 5 to 1

The shadow price associated with the second constraint (c2) is 1.333, which means that for every one-unit increase in the RHS of constraint c2, the optimal objective function value increases by 1.3333.

Now, if the coefficient for x2 in constraint c2 changes from 5 to 1, it would impact the RHS of c2. The Shadow Price indicates that the optimal objective function value would increase by 1.2 for each unit increase in the RHS of c2.

Since the Shadow Price (1.2) is greater than the reduced cost (RC) of x2 (0), it suggests that increasing the RHS of c2 (and thereby decreasing the coefficient for x2) could improve the objective function value. Therefore, the current solution may no longer be optimal if the coefficient for x2 in the second constraint changes from 5 to 1.

# Question 2

In [5]:
import gurobipy as gp
import pandas as pd

In [20]:
cards, profit = gp.multidict({
    'HyperLink' : 53,
    'FastLink'  : 48,
    'SpeedLink' : 33,
    'MicroLink' : 32,
    'EtherLink' : 38
})

devices, devices_lower_bound, devices_upper_bound = gp.multidict({
    'Circuit'  :  [0, 80000],
    'Resistors':  [0, 100000],
    'Memory'   :  [0, 30000],
    'Labor'    :  [0, 5000]
})

device_value = {
    ('HyperLink', 'Circuit'): 20,
    ('HyperLink', 'Resistors'): 28,
    ('HyperLink', 'Memory'): 8,
    ('HyperLink', 'Labor'): 0.75,
    ('FastLink', 'Circuit'): 15,
    ('FastLink', 'Resistors'): 24,
    ('FastLink', 'Memory'): 8,
    ('FastLink', 'Labor'): 0.6,
    ('SpeedLink', 'Circuit'): 10,
    ('SpeedLink', 'Resistors'): 18,
    ('SpeedLink', 'Memory'): 4,
    ('SpeedLink', 'Labor'): 0.5,
    ('MicroLink', 'Circuit'): 8,
    ('MicroLink', 'Resistors'): 12,
    ('MicroLink', 'Memory'): 4,
    ('MicroLink', 'Labor'): 0.65,
    ('EtherLink', 'Circuit'): 5,
    ('EtherLink', 'Resistors'): 16,
    ('EtherLink', 'Memory'): 6,
    ('EtherLink', 'Labor'): 1
}

In [24]:
Link = Model('Interface cards')
number = Link.addVars(cards, name="number")
Link.setObjective(number.prod(profit), GRB.MAXIMIZE)
Link.addConstrs((gp.quicksum(device_value[c, d] * number[c] for c in cards)\
                 == [devices_lower_bound[d], devices_upper_bound[d]]\
                      for d in devices), "_")
# Add the constraint that each variable must be greater than 500
Link.addConstrs(number[c] >= 500 for c in cards)

Link.addConstr(2 * number['HyperLink'] <= number['FastLink'])
Link.write('InterfaceCards.lp')

def printSolution():
    if Link.status == GRB.OPTIMAL:
        print('\nProfit: %g' % Link.objVal)
        print('\nnumber:')
        for c in cards:
            if number[c].x > 0.0001:
                print('%s %g' % (c, number[c].x))
    else:
        print('No solution')

# Solve
Link.optimize()
printSolution()

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 10 rows, 9 columns and 31 nonzeros
Model fingerprint: 0xc30dd8ff
Coefficient statistics:
  Matrix range     [5e-01, 3e+01]
  Objective range  [3e+01, 5e+01]
  Bounds range     [5e+03, 1e+05]
  RHS range        [5e+02, 1e+05]
Presolve removed 5 rows and 4 columns
Presolve time: 0.00s
Presolved: 5 rows, 5 columns, 22 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    4.5660000e+05   1.167975e+04   0.000000e+00      0s
       3    2.1500000e+05   0.000000e+00   0.000000e+00      0s

Solved in 3 iterations and 0.01 seconds (0.00 work units)
Optimal objective  2.150000000e+05

Profit: 215000

number:
HyperLink 500
FastLink 1000
SpeedLink 1500
MicroLink 2250
EtherLink 500


In [25]:
Link.ObjVal

215000.0

## A. Use Gurobi API to create a Sensitivity Report

In [26]:
print('Sensitivity Analysis (SA) \n ObjVal =', Link.ObjVal)
Link.printAttr(['X', 'RC', 'Obj', 'SAObjLow', 'SAObjUp'])
Link.printAttr(['Sense', 'Slack', 'Pi', 'RHS', 'SARHSLow', 'SARHSUp'])

Sensitivity Analysis (SA) 
 ObjVal = 215000.0

    Variable            X           RC          Obj     SAObjLow      SAObjUp 
-----------------------------------------------------------------------------
number[HyperLink]          500            0           53         -inf      96.6667 
number[FastLink]         1000            0           48         -inf           64 
number[SpeedLink]         1500            0           33           32           48 
number[MicroLink]         2250            0           32      26.7273           33 
number[EtherLink]          500            0           38         -inf      47.6667 
   RgCircuit        19500            0            0        -65.5          0.5 
 RgResistors            0    -0.166667            0         -inf     0.166667 
    RgMemory            0         -7.5            0         -inf          7.5 
     RgLabor       1312.5            0            0     -6.66667      22.8571 

  Constraint        Sense        Slack           Pi         

## B. Which of the constraints in the problem are binding?

The constraint with name: Resistors, Memory, R4, R8, R9

## C.  If the company was going to eliminate one of its products, which one should it be?

Microlink 

## D.  If the company could by 1,000 additional memory chips at the usual cost, should they, do it? If so, how much would profits increase? 

Yes, they should do it because the profit increase $7500 from 215000 to 222500

## E. Suppose the manufacturing costs used in this analysis were estimated hastily and are known to be somewhat imprecise. Which products would you be most concerned about having more precise cost estimates for before implementing this solution?

 hyperlink, fastlink, etherlink

# Question 3

In [31]:
supply_locs, supply = gp.multidict({
    'Location 1': 16,
    'Location 2': 18
    })

demand_locs, demand_lower, demand_upper = gp.multidict({
    'Location 3': [5,10],
    'Location 4': [5,10],
    'Location 5': [5,10],
    'Location 6': [5,10]
    })

arcs, cost = gp.multidict({
    ('Location 1', 'Location 3'): 54,
    ('Location 1', 'Location 4'): 17,
    ('Location 1', 'Location 5'): 23,
    ('Location 1', 'Location 6'): 30,
    ('Location 2', 'Location 3'): 24,
    ('Location 2', 'Location 4'): 18,
    ('Location 2', 'Location 5'): 19,
    ('Location 2', 'Location 6'): 31,
})

In [32]:
car = gp.Model('Transshipment')
flow = car.addVars(arcs, obj=cost, name="flow")

demand_flow = car.addConstrs((gp.quicksum(flow.select("*", d_loc)) \
                             == [demand_lower[d_loc], demand_upper[d_loc]]\
                                for d_loc in demand_locs), "_") 

supply_flow = car.addConstrs((gp.quicksum(flow.select(s_loc, '*')) \
                               == supply[s_loc]\
                                for s_loc in supply_locs), "_") 

car.optimize()

Gurobi Optimizer version 11.0.1 build v11.0.1rc0 (win64 - Windows 10.0 (19045.2))

CPU model: AMD Ryzen 9 5900HS with Radeon Graphics, instruction set [SSE2|AVX|AVX2]
Thread count: 8 physical cores, 16 logical processors, using up to 16 threads

Optimize a model with 6 rows, 12 columns and 20 nonzeros
Model fingerprint: 0x9bdc6bc1
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [2e+01, 5e+01]
  Bounds range     [5e+00, 5e+00]
  RHS range        [1e+01, 2e+01]
Presolve time: 0.00s
Presolved: 6 rows, 12 columns, 20 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   4.400000e+01   0.000000e+00      0s
       7    7.3000000e+02   0.000000e+00   0.000000e+00      0s

Solved in 7 iterations and 0.00 seconds (0.00 work units)
Optimal objective  7.300000000e+02


## A. Use Gurobi API to create a Sensitivity Report

In [33]:
print('Sensitivity Analysis (SA) \n ObjVal =', car.ObjVal)
car.printAttr(['X', 'RC', 'Obj', 'SAObjLow', 'SAObjUp'])
car.printAttr(['Sense', 'Slack', 'Pi', 'RHS', 'SARHSLow', 'SARHSUp'])
# Note: printAttr prints only rows with at least one NON-ZERO value, e.g. m.printAttr('X') prints only non-zero variable values.

Sensitivity Analysis (SA) 
 ObjVal = 730.0

    Variable            X           RC          Obj     SAObjLow      SAObjUp 
-----------------------------------------------------------------------------
flow[Location 1,Location 3]            0           26           54           28          inf 
flow[Location 1,Location 4]           10            0           17         -inf           22 
flow[Location 1,Location 5]            1            0           23           18           25 
flow[Location 1,Location 6]            5            0           30           28           35 
flow[Location 2,Location 3]            9            0           24           19           26 
flow[Location 2,Location 4]            0            5           18           13          inf 
flow[Location 2,Location 5]            9            0           19           17           24 
flow[Location 2,Location 6]            0            5           31           26          inf 
RgLocation 3            1            0         

In [ ]:
# D
# Add constraint sending 1 car from L1 to L3
#car.addConstr(flow['Location 1', 'Location 3'] == 1")

In [ ]:
# E
# Add constraint for location 6 demand
car.addConstr(flow.sum('*', 'Location 6') >= 8, name="demand_L6")

B. Not unique because RC are not all 0

C. Location 6 because it receives 5 cars from L1 and 0 car from l2 which is 5 cars in total 

D. Cost increases from 730 to 756 => 26

E. Cost increases from 730 to 746 => 16